cpp/src_prims/sparse/distance/l2_distance.cuh

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <limits.h>
+#include <cmath>
+
+#include <raft/cudart_utils.h>
+#include <sparse/distance/common.h>
+
+#include <raft/cudart_utils.h>
+#include <raft/linalg/distance_type.h>
+#include <raft/sparse/cusparse_wrappers.h>
+#include <raft/cuda_utils.cuh>
+#include <raft/linalg/unary_op.cuh>
+
+#include <raft/mr/device/allocator.hpp>
+#include <raft/mr/device/buffer.hpp>
+
+#include <sparse/utils.h>
+#include <sparse/csr.cuh>
+
+#include <sparse/distance/common.h>
+#include <sparse/distance/ip_distance.cuh>
+
+#include <cuml/neighbors/knn.hpp>
+
+#include <nvfunctional>
+
+namespace raft {
+namespace sparse {
+namespace distance {
+
+// @TODO: Move this into sparse prims (coo_norm)
+template <typename value_idx, typename value_t>
+__global__ void compute_row_norm_kernel(value_t *out, const value_idx *coo_rows,
+                                        const value_t *data, value_idx nnz) {
+  value_idx i = blockDim.x * blockIdx.x + threadIdx.x;
+  if (i < nnz) {
+    atomicAdd(&out[coo_rows[i]], data[i] * data[i]);
```

Comment:
Good idea! I will add this to the issues with further optimizations

Reply:
Thinking about this more closely- I recall that since the cusparse gemm required B to be transposed, I am using B's indices array to compute the row norms, which are going to be sorted by column. Thinking about this further, though, I could just use the original CSR indptr arrays and do warp-level (and potentially block-level if they get large enough) reductions for each row.